### ISSUE TITLE
Investigate roadblocks to agile LCM development

### ISSUE THREAD

**Issue by jtostie** [gahansen/Albany#22]
1. Proliferation of template parameters inhibits non experts from development
2. How to manage code complexity?

**Comment by mjuha** [comment]
From my perspective, one of the problem is that the data encapsulation concept of C++ is not enforced. For new implementations we should use this concept. Also, we need to use more comments in the interfaces. Ideally I should not need to go to the source code to figure out what the function is doing.

**Comment by lxmota** [comment]
Yes, this is a big one. At the recent Albany Users Meeting it was decided that @ibaned would take some of the work in maintaining Albany's infrastructure, which is a very important issue for LCM.

Besides that, there are other issues too, such as the steep learning curve, the horrifying XML of the input files and others.

In my experience, many times the two issues mentioned by @jtostie above arise because code is just copied over and over and modified slightly to add new features. This is a complex problem. There are some ongoing efforts by myself and others (@calleman21 , @jtostie) to try to address this. But it will take time and a lot of work.
